codex/stores/sqlitestore.nim

```diff
@@ -0,0 +1,204 @@
+  T: type SQLiteStore,
+  repoDir: string,
+  cache: BlockStore = CacheStore.new()): T =
+
+  T(cache: cache,
+    datastore: SQLiteDatastore.new(repoDir).expect("should not fail"))
+
+proc datastore*(self: SQLiteStore): SQLiteDatastore =
+  self.datastore
+
+proc blockKey*(blockCid: Cid): Key =
+  Key.init($blockCid).expect("should not fail")
+
+method getBlock*(
+  self: SQLiteStore,
+  cid: Cid): Future[?!(?Block)] {.async.} =
+  ## Get a block from the cache or database, save a copy to the cache
+  ##
+
+  trace "Getting block from database", cid
+  if cid.isEmpty:
+    trace "Empty block, ignoring"
+    return success cid.emptyBlock.some
+
+  let
+    getCachedRes = await self.cache.getBlock(cid)
+
+  if getCachedRes.isErr:
+    trace "Unable to read block from cache", cid, error = getCachedRes.error.msg
+
+  if getCachedRes.get.isSome:
+    trace "Retrieved block from cache", cid
+    return getCachedRes
+
+  let
+    blkKey = blockKey(cid)
+    getRes = await self.datastore.get(blkKey)
+
+  if getRes.isErr:
+    trace "Unable to read block", key = blkKey.id, error = getRes.error.msg
+    return failure "Unable to read block"
+
+  let
+    dataOpt = getRes.get
+
+  if dataOpt.isNone:
+    return success Block.none
+
+  without blk =? Block.new(cid, dataOpt.get), error:
+    trace "Unable to construct block from data", cid, error = error.msg
+    return failure error
+
+  let
+    putCachedRes = await self.cache.putBlock(blk)
+
+  if putCachedRes.isErr:
+    trace "Unable to store block in cache", cid, error = putCachedRes.error.msg
+
+  return success blk.some
+
+method putBlock*(
+  self: SQLiteStore,
+  blk: Block): Future[?!void] {.async.} =
+  ## Write a block's contents to the database with key based on blk.cid,
+  ## save a copy to the cache
+  ##
+
+  if blk.isEmpty:
+    trace "Empty block, ignoring"
+    return success()
+
+  let
+    cid = blk.cid
+    blkKey = blockKey(cid)
+    putRes = await self.datastore.put(blkKey, blk.data)
+
+  if putRes.isErr:
+    trace "Unable to store block", key = blkKey.id, error = putRes.error.msg
+    return failure "Unable to store block"
+
+  let
+    putCachedRes = await self.cache.putBlock(blk)
+
+  if putCachedRes.isErr:
+    trace "Unable to store block in cache", cid, error = putCachedRes.error.msg
+
+  return success()
+
+method delBlock*(
+  self: SQLiteStore,
+  cid: Cid): Future[?!void] {.async.} =
+  ## Delete a block from the database and cache
+  ##
+
+  trace "Deleting block from database", cid
+  if cid.isEmpty:
+    trace "Empty block, ignoring"
+    return success()
+
+  let
+    blkKey = blockKey(cid)
+    delRes = await self.datastore.delete(blkKey)
+
+  if delRes.isErr:
+    trace "Unable to delete block", key = blkKey.id, error = delRes.error.msg
+    return failure "Unable to delete block"
+
+  let
+    delCachedRes = await self.cache.delBlock(cid)
+
+  if delCachedRes.isErr:
+    trace "Unable to delete block from cache", cid,
+      error = delCachedRes.error.msg
```

Comment:
Here I also only trace, but `FSSstore` returns the failure if there was an error deleting from the cache, which is at odds with the behavior of `putBlock` for `FSStore`

https://github.com/status-im/nim-codex/blob/7cfb157d327624d848a09c2df4eda20ea677a8e4/codex/stores/fsstore.nim#L124

Reply:
We probably want the same behaviour in FSStore either way